File: nestjs-module/{{cookiecutter.fileName}}/service/{{cookiecutter.fileName}}.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { CollectionDto } from '@sigmaott/paginate';
import { I18nContext } from 'nestjs-i18n';

import { Create{{cookiecutter.className}}Dto } from '../dto/create-{{cookiecutter.fileName}}.dto';
import { Update{{cookiecutter.className}}Dto } from '../dto/update-{{cookiecutter.fileName}}.dto';
import { {{cookiecutter.className}}, {{cookiecutter.className}}Collection } from '../entities/{{cookiecutter.fileName}}.entity';
import { {{cookiecutter.className}}Repository } from '../repository/{{cookiecutter.fileName}}.repository';

@Injectable()
export class {{cookiecutter.className}}Service {
  constructor(private readonly {{cookiecutter.fieldPrefix}}Repository: {{cookiecutter.className}}Repository) { }

  async create(
    appId: string,
    payload: Create{{cookiecutter.className}}Dto,
    i18n: I18nContext
  ) {
    
    const modelCreate = {
      ...payload,
      appId:appId
    } as {{cookiecutter.className}};

   
    const res = await this.{{cookiecutter.fieldPrefix}}Repository.create(modelCreate);

    return res;
  }

  async findAll(
    appId: string,
    collectionDto: CollectionDto
  ): Promise< {{cookiecutter.className}}Collection> {
    collectionDto.filter.appId = appId

    return this.{{cookiecutter.fieldPrefix}}Repository.paginate(collectionDto);
  }

  async findOne(id: string, appId: string, i18n: I18nContext) {
   

    const res = await this.{{cookiecutter.fieldPrefix}}Repository.findOne({ id: id});
    if (!res||res.appId !== appId){
      return 
    }
    return res
  }


  async updateOne(
    id: string,
    appId: string,
    payload: Update{{cookiecutter.className}}Dto,
    i18n: I18nContext
  ) {
 

    return await this.{{cookiecutter.fieldPrefix}}Repository.updateOne(
      { id: id, appId: appId },
      payload
    );
  }

  async deleteOne(appId: string, id: string, i18n: I18nContext) {

    const res = await this.{{cookiecutter.fieldPrefix}}Repository.findOne({ id: id });
    if (!res||res.appId !== appId){
      return 
    }

    return await this.{{cookiecutter.fieldPrefix}}Repository.deleteOne({ id: id });


  }

  




}

File: fx-tdm-module/{{cookiecutter.module}}/service/{{cookiecutter.fileName}}_service.go
package service

import (
	"context"
	"{{cookiecutter.basePath}}/dto"
	"{{cookiecutter.basePath}}/model"
	"{{cookiecutter.basePath}}/repository"

	"github.com/google/uuid"
	base "github.com/sigmaott/gest/package/technique/database/base"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
)

type I{{cookiecutter.className}}Service interface {
	Create(payload *dto.Create{{cookiecutter.className}}, appId string) (*model.{{cookiecutter.className}}, error)
	FindOne(params struct {
		Id    string
		AppId string
	}) (*model.{{cookiecutter.className}}, error)
	Paginate(query any, paginate *base.Paginate, sort *base.Sort) (results *base.PaginateResponse[model.{{cookiecutter.className}}], err error)
	UpdateOne(params struct {
		Id    string
		AppId string
	}, payload *dto.Update{{cookiecutter.className}}) (*model.{{cookiecutter.className}}, error)
	DeleteOne(params struct {
		Id    string
		AppId string
	}) (*model.{{cookiecutter.className}}, error)
}
type {{cookiecutter.fieldPrefix}}Service struct {
	{{cookiecutter.fieldPrefix}}Repository repository.I{{cookiecutter.className}}Repository
}

func (s *{{cookiecutter.fieldPrefix}}Service) Create(payload *dto.Create{{cookiecutter.className}}, appId string) (*model.{{cookiecutter.className}}, error) {

	modelCreated := &model.{{cookiecutter.className}}{
		Id: uuid.New().String(),
	}
	res, err := s.{{cookiecutter.fieldPrefix}}Repository.CreateOne(context.TODO(), modelCreated)
	return res, err
}

func (s *{{cookiecutter.fieldPrefix}}Service) FindOne(params struct {
	Id    string
	AppId string
}) (*model.{{cookiecutter.className}}, error) {
	res, err := s.{{cookiecutter.fieldPrefix}}Repository.FindOne(context.TODO(), bson.M{
		"id": params.Id,
	})
	if err != nil {
		return res, nil
	}
	if res.AppId != params.AppId {
		return nil, mongo.ErrNilDocument
	}
	return res, err
}

// UpdateOne implements IExampleService.
func (s *{{cookiecutter.fieldPrefix}}Service) UpdateOne(params struct {
	Id    string
	AppId string
}, payload *dto.Update{{cookiecutter.className}}) (*model.{{cookiecutter.className}}, error) {
	res, err := s.{{cookiecutter.fieldPrefix}}Repository.FindOne(context.TODO(), bson.M{
		"id": params.Id,
	})
	if err != nil {
		return res, nil
	}
	if res.AppId != params.AppId {
		return nil, mongo.ErrNilDocument
	}

	res, err = s.{{cookiecutter.fieldPrefix}}Repository.UpdateOne(context.TODO(), bson.M{
		"id": params.Id,
	}, payload)
	return res, err
}

func (s *{{cookiecutter.fieldPrefix}}Service) DeleteOne(params struct {
	Id    string
	AppId string
}) (*model.{{cookiecutter.className}}, error) {
	res, err := s.{{cookiecutter.fieldPrefix}}Repository.FindOne(context.TODO(), bson.M{
		"id": params.Id,
	})
	if err != nil {
		return res, nil
	}
	if res.AppId != params.AppId {
		return nil, mongo.ErrNilDocument
	}

	res, err = s.{{cookiecutter.fieldPrefix}}Repository.DeleteOne(context.TODO(), bson.M{
		"id": params.Id,
	})
	return res, err
}

// Paginate implements IExampleService.
func (s *{{cookiecutter.fieldPrefix}}Service) Paginate(query any, paginate *base.Paginate, sort *base.Sort) (results *base.PaginateResponse[model.{{cookiecutter.className}}], err error) {
	return s.{{cookiecutter.fieldPrefix}}Repository.Paginate(context.TODO(), query, paginate, sort)
}

func New{{cookiecutter.className}}Service(exampleRepository repository.I{{cookiecutter.className}}Repository) I{{cookiecutter.className}}Service {

	return &{{cookiecutter.fieldPrefix}}Service{
		{{cookiecutter.fieldPrefix}}Repository: {{cookiecutter.fieldPrefix}}Repository,
	}

}
